CMakeLists.txt

```diff
@@ -124,98 +126,43 @@ else()
+endfunction()
 
-# ---------------------------------------------------------------------------------------
-# This insanity is for people embedding libavif or making fully static or Windows builds.
-# Any proper unix environment should ignore these entire following blocks.
 option(AVIF_LOCAL_ZLIBPNG "Build zlib and libpng by providing your own copy inside the ext subdir." OFF)
 if(AVIF_LOCAL_ZLIBPNG)
-    # Put the value of ZLIB_INCLUDE_DIR in the cache. This works around cmake behavior that has been updated by
-    # cmake policy CMP0102 in cmake 3.17. Remove the CACHE workaround when we require cmake 3.17 or later. See
-    # https://gitlab.kitware.com/cmake/cmake/-/issues/21343.
-    set(ZLIB_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ext/zlib" CACHE PATH "zlib include dir")
-    # This include_directories() call must be before add_subdirectory(ext/zlib) to work around the
-    # zlib/CMakeLists.txt bug fixed by https://github.com/madler/zlib/pull/818.
-    include_directories(SYSTEM ${ZLIB_INCLUDE_DIR})
-    add_subdirectory(ext/zlib)
-    # This include_directories() call and the previous include_directories() call provide the zlib
-    # include directories for add_subdirectory(ext/libpng). Because we set PNG_BUILD_ZLIB,
-    # libpng/CMakeLists.txt won't call find_package(ZLIB REQUIRED) and will see an empty
-    # ${ZLIB_INCLUDE_DIRS}.
-    include_directories("${CMAKE_CURRENT_BINARY_DIR}/ext/zlib")
-    set(CMAKE_DEBUG_POSTFIX "")
-
-    # This is the only way I could avoid libpng going crazy if it found awk.exe, seems benign otherwise
-    set(PREV_ANDROID ${ANDROID})
-    set(ANDROID TRUE)
-    set(PNG_BUILD_ZLIB "${CMAKE_CURRENT_SOURCE_DIR}/ext/zlib" CACHE STRING "" FORCE)
-    set(PNG_SHARED OFF CACHE BOOL "")
-    set(PNG_TESTS OFF CACHE BOOL "")
-    add_subdirectory(ext/libpng)
-    set(PNG_PNG_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ext/libpng")
-    set(PNG_LIBRARY png_static)
-    include_directories("${CMAKE_CURRENT_BINARY_DIR}/ext/libpng")
-    set(ANDROID ${PREV_ANDROID})
-
-    set(ZLIB_LIBRARY zlibstatic)
+    include(LocalZlibpng)
 endif()
 
 option(AVIF_LOCAL_JPEG "Build jpeg by providing your own copy inside the ext subdir." OFF)
 if(AVIF_LOCAL_JPEG)
-    add_subdirectory(ext/libjpeg)
-    if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
-        set(JPEG_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ext/libjpeg")
-        set(JPEG_LIBRARY jpeg)
-    else()
-        set(JPEG_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ext/libjpeg" PARENT_SCOPE)
-        set(JPEG_LIBRARY jpeg PARENT_SCOPE)
-    endif()
+    include(LocalJpeg)
 endif()
 
 set_local_or_system_option("LIBYUV" "OFF" "Use libyuv.")
-if(AVIF_LIBYUV STREQUAL "LOCAL")
-    set(AVIF_LIBYUV_BUILD_DIR "${CMAKE_CURRENT_SOURCE_DIR}/ext/libyuv/build")
-
-    # If ${ANDROID_ABI} is set, look for the library under that subdirectory.
-    if(DEFINED ANDROID_ABI)
-        set(AVIF_LIBYUV_BUILD_DIR "${AVIF_LIBYUV_BUILD_DIR}/${ANDROID_ABI}")
+if(NOT TARGET yuv)
```

Comment:
IMPORTANT: Actually, not all 3rd party deps behave the same. E.g., `yuv` checks for the presence of the target before checking for `LOCAL` and `SYSTEM`. Not libsharpyuv.
I propose for a 3rd party dep:
- if `OFF`: do not use it
- if `LOCAL`: use the one in ext
- if `SYSTEM`: use whatever is already here (if built by the outer CMake project), otherwise, find_package it REQUIRED

For variables with only two choices (ON/OFF), e.g. AVIF_LOCAL_LIBXML2:
- ON: use the one in ext
- OFF: use whatever is already here (if built by the outer CMake project), otherwise, find_package it REQUIRED

Reply:
I think we might not have a choice: if the target already exists in the outer project, a second attempt to call `add_library` with that same target name would fail.

Reply:
Right, so we should always check right? E.g.
```
if(NOT TARGET dav1d::dav1d)
```

Reply:
Yes, that makes sense. I thought your second and third bullet points were proposing that we override an outer target with the one in ext if LOCAL is passed, but I think we can reasonably interpret any value in the context of an outer project as using the outer project's target

Reply:
Just to be clear, I suggest:
```
if(AVIF_CODEC_FOO)
  if(AVIF_CODEC_FOO STREQUAL "LOCAL")
    # use the version in ext no matter what
  else()
    if(TARGET foo::foo)
      # Use foo that is built by the outer CMake because it is as if it were on the system
    else()
      find_package(foo REQUIRED)
    endif()
  endif()
else()
  # do not use codec foo no matter what 
endif()
```
And for the two option version
```
if(AVIF_LOCAL_FOO)
  # use the version in ext no matter what
else()
  if(TARGET foo::foo)
    # Use foo that is built by the outer CMake because it is as if it were on the system
  else()
    find_package(foo REQUIRED)
  endif()
endif()
```

Reply:
Okay, that's what I thought you meant. The trouble with that is that if the target already exists, you can't create a different target with that name. If a parent project has the target because it called find_package or add_subdirectory, or because it created an externalproject target, there isn't really anything we can do about it. Even if it were possible to work around this limitation, I don't think you could use a different build of the third party dependency in libavif without symbol conflicts. So we could either use that target and assume that the parent project knows what it's doing, or else throw an error if LOCAL is used when there is already a target.

Reply:
It might well be the case that the parent project's targets are local to the build, so requiring SYSTEM for that use-case could be confusing. But if the error message is sufficiently clear I think that's fine